map/autoware_pointcloud_divider/src/pcd_divider.cpp

```diff
@@ -0,0 +1,482 @@
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <autoware/pointcloud_divider/pcd_divider.hpp>
+#include <autoware/pointcloud_divider/utility.hpp>
+#include <autoware/pointcloud_divider/voxel_grid_filter.hpp>
+
+#include <pcl/common/transforms.h>
+#include <pcl/filters/voxel_grid.h>
+
+#include <filesystem>
+#include <memory>
+
+namespace fs = std::filesystem;
+
+namespace autoware::pointcloud_divider
+{
+
+template <class PointT>
+std::vector<std::string> PCDDivider<PointT>::discoverPCDs(const std::string & input)
+{
+  // Discover PCD files in the input_dir
+  std::vector<std::string> pcd_list;
+  fs::path input_path(input);
+
+  if (fs::is_directory(input_path)) {
+    RCLCPP_INFO(logger_, "Input PCD directory: %s", input.c_str());
+
+    for (auto & entry : fs::directory_iterator(input_path)) {
+      if (fs::is_regular_file(entry.symlink_status())) {
+        auto file_name = entry.path().string();
+        auto extension = entry.path().extension().string();
+
+        if (extension == ".pcd" || extension == ".PCD") {
+          pcd_list.push_back(file_name);
+        }
+      }
+    }
+  } else if (fs::is_regular_file(input_path)) {
+    auto file_name = input_path.string();
+    auto extension = input_path.extension().string();
+
+    if (extension == ".pcd" || extension == ".PCD") {
+      RCLCPP_INFO(logger_, "Input PCD file: %s", input.c_str());
+
+      pcd_list.push_back(file_name);
+    } else {
+      RCLCPP_ERROR(logger_, "Error: The input file is not PCD format %s", input.c_str());
+      exit(EXIT_FAILURE);
+    }
+  } else {
+    RCLCPP_ERROR(logger_, "Error: Invalid input %s", input.c_str());
+    exit(EXIT_FAILURE);
+  }
+
+  return pcd_list;
+}
+
+template <class PointT>
+void PCDDivider<PointT>::run()
+{
+  // Discover PCD files
+  auto pcd_list = discoverPCDs(input_pcd_or_dir_);
+
+  // Process pcd files
+  run(pcd_list);
+}
+
+template <class PointT>
+void PCDDivider<PointT>::run(const std::vector<std::string> & pcd_names)
+{
+  checkOutputDirectoryValidity();
+
+  grid_set_.clear();
+
+  for (const std::string & pcd_name : pcd_names) {
+    if (!rclcpp::ok()) {
+      return;
+    }
+
+    if (debug_mode_) {
+      RCLCPP_INFO(logger_, "Dividing file %s", pcd_name.c_str());
+    }
+
+    do {
+      auto cloud_ptr = loadPCD(pcd_name);
+
+      dividePointCloud(cloud_ptr);
+    } while (reader_.good() && rclcpp::ok());
+  }
+
+  saveTheRest();
+
+  RCLCPP_INFO(logger_, "Merge and downsampling... ");
+
+  // Now merge and downsample
+  mergeAndDownsample();
+
+  std::string yaml_file_path = output_dir_ + "/" + file_prefix_ + "_metadata.yaml";
```

Comment:
```suggestion
  std::string yaml_file_path = output_dir_ + "/" + "pointcloud_map_metadata.yaml";
```

The metadata file name is recommended to be "pointcloud_map_metadata.yaml," so I would like to ensure it is always saved with that name.

https://github.com/autowarefoundation/autoware.universe/tree/main/map/map_loader#directory-structure-of-these-files